contributors/design-proposals/device-plugin.md

```diff
@@ -0,0 +1,553 @@
+3. The device plugin start it's gRPC server if it did not recieve an
+   error
+
+## Unix Socket
+
+Device Plugins are expected to communicate with Kubelet through gRPC
+on an Unix socket.
+When starting the gRPC server, they are expected to create a unix socket
+at the following host path: `/var/run/kubernetes`.
+
+For non bare metal device plugin this means they will have to mount the folder
+as a volume in their pod spec ([see Installation](##installation)).
+
+Device plugins can expect to find the socket to register themselves on
+the host at the following path:
+`/var/run/Kubernetes/kubelet.sock`.
+
+## Protocol Overview
+
+When first registering themselves against Kubelet, the device plugin
+will send:
+  * The name of their unix socket
+  * [The API version against which they were built](#versioning).
+  * Their `ResourceName` they want to advertise
+
+Kubelet answers with the minimum version it supports and whether or
+not there was an error. The errors may include (but not limited to):
+  * API version not supported
+  * A device plugin already registered this `ResourceName`
+
+Kubelet will then interact with the plugin through the following functions:
+  * ListAndWatch: The device plugin advertises a list of Devices to Kubelet
+    and sends it again if the state of a Device changes
+  * `Allocate`: Called when creating a container with a list of devices
+
+![Process](device-plugin.png)
+
+
+## API Specification
+
+```go
+// Registration is the service advertised by the Kubelet
+// Only when Kubelet answers with a success code to a Register Request
+// may Device Plugins start their service
+service Registration {
+	rpc Register(RegisterRequest) returns (RegisterResponse) {}
+}
+
+// DevicePlugin is the service advertised by Device Plugins
+service DevicePlugin {
+	// ListWatch returns a stream of List of Devices
+	// Whenever a Device state change or a Device disapears, ListWatch
+	// returns the new list
+	rpc ListWatch(Empty) returns (stream ListWatchResponse) {}
+
+	// Allocate is called during container creation so that the Device
+	// Plugin can run device specific operations and instruct Kubelet
+	// of the steps to make the Device available in the container
+	rpc Allocate(AllocateRequest) returns (AllocateResponse) {}
+}
+
+message RegisterRequest {
+	// Version of the API the Device Plugin was built against
+	string version = 1;
+	// Name of the unix socket the device plugin is listening on
+	string endpoint = 2;
+	// Name of the Device advertised same as the one in the Device struct
+	string resourceName = 3;
+}
+
+message RegisterResponse {
+	// Minimum version the Kubelet API supports.
+	bool success = 1;
+	// Kubelet fills this field if it encounters any errors
+	// during the registration process
+	string error = 2;
+}
+
+// Allocate is called during container creation so that the Device
+// Plugin can run device specific operations on the Devices sent
+message AllocateRequest {
+	repeated Device devices = 1;
```

Comment:
+1 to what @vishh said and just re-stated the reasons why I hope we can just use simple device ID here. Generally, an API serves as a contract between the two sides of parties using the API. We would like to keep APIs concise because otherwise, people may get confused when looking at or using the API. E.g., if we use the whole Device struct here, a future device plugin developer may wonder whether all the fields in the Device struct needs to be validated before serving an Allocate request, or whether they convey extra information during allocation. This could lead to more diverse implementations of device plugin and be more error prone. This API might be convenient for the current prototype implementation but it is going to be used and maintained by other future developers on device plugin or Kubelet. That is why we hope we can keep it simple and clear.